### PR TITLE
Implicit flow is too implicitly described

### PR DESCRIPTION
**Pull request by Riduidel**
It took me two weeks to have a working implicit OpenID Connect flow. My main errors were elements not sufficiently described (well, to my mind). I complained a lot at the lack of documentation, and would prefer not to fall in that trap again. So i tried to improve documentation on that very subject.

### CODE REVIEW

File: md-docs/_1x/guides/authentication/openid/overview.md [hunk 1]
```diff
@@ -142,12 +142,12 @@ Wait for the web view to redirect to a URL whose host and path are the same as t
...
+2. Client includes the ID token (as a bearer token on the `Authorization` header) on requests made against the Sync Gateway REST API. 
 3. Sync Gateway matches the token to a provider in its config based on the issuer and audience in the tokwn.
 4. Sync Gateway validates the token, based on the provider definition.
 5. On successful validation, Sync Gateway authenticates the user based on the subject and issuer in the token.
 
-Since ID tokens are typically large, the usual approach is to use the ID token to obtain a Sync Gateway session (using the /db/_session REST endpoint), and then use that token for subsequent authentication requests.
+Since ID tokens are typically large, the usual approach is to use the ID token to obtain a Sync Gateway session (using the `/db/_session` REST endpoint), and then use that token for subsequent authentication requests. During this call, the token will be sent as `Authorization` header with the `Bearer ` prefix. Notice that only tokens signed are supported, which can be verified using [jwt.io Debugger](https://jwt.io/#debugger-io). in this debugger, such tokens have the `alg` set to `RS256`, and there is a `kid` element in the header.
```
Comment (jamiltz) [inline md-docs/_1x/guides/authentication/openid/overview.md]: > Notice that only tokens signed are supported, which can be verified using [jwt.io Debugger](https://jwt.io/#debugger-io)

Nice tool! Could you confirm the flow you went through to test authentication was the following:

1. Get ID token from provider
2. Verify token is `RS256` with https://jwt.io/#debugger-io
3. Send token in `/{db}/_session` request to SG
4. Use returned session ID for subsequent requests

Out of curiosity, did you use the test provider at all? How are you generating tokens in step 1 for testing purposes?

Comment (jamiltz) [inline md-docs/_1x/guides/authentication/openid/overview.md]: Basically, I'm curious to know if there's a way to test the implicit flow using an external provider (similar to the jwt.io debugger but for the full implicit flow). Right now, this example is tied to the Google SignIn API.

**Comment by jamiltz** [comment]
related forum thread https://forums.couchbase.com/t/cant-initiate-an-implicit-openid-connect-flow/12950/5

File: md-docs/_1x/guides/authentication/openid/overview.md (hunk 1 above)
Comment (Riduidel) [inline md-docs/_1x/guides/authentication/openid/overview.md]: @jamiltz The tests I used was indeed the ones you mentionned, in the order given there.

I didn't use the test provider, as it seems to me the behaviour is quite different (typically, the test provider has specific code in auth.go to handle its code, which makes it a non-representative for integration tests like OpenID, and so on.

The provider I used is auth0, which provides a free test plan of 30 days, of which I used only 5 to perform my tests. And, to be complete, I used a JavaFX example based upon the couchbase lite sample (https://blog.couchbase.com/using-couchbase-in-a-javafx-desktop-application/) as a start point. I must confess inserting the Auth0 WebView was not that easy, but it is more due to my will to reinject a little CDI in.

Comment (adamcfraser) [inline md-docs/_1x/guides/authentication/openid/overview.md]: @Riduidel When you say "Notice that only tokens signed are supported...", what's the context?  I believe that Sync Gateway enforces the supported signing algorithms based on what's published by your OpenID Connect provider (via the discovery endpoint), and not based on anything internal to SG.

**Comment by jamiltz** [comment]
@Riduidel I've merged your changes and re-arranged them a bit (the reference to the debugger is particularly useful) https://github.com/couchbaselabs/couchbase-mobile-portal/commit/8fd62f3ef942c9a4884a5627677ce9a99b868e5d -
 https://developer.couchbase.com/documentation/mobile/1.4/guides/authentication/openid/index.html#implicit-flow

I've also emphasized on the fact that Sync Gateway supports both `HS256` and `RS256`. If this has not been the case in your testing can you let us know what worked/didn't work?

**Comment by Riduidel** [comment]
Sorry for the late reply, but from my tests, the HS256 never appeared as supported.
(I know @adamcfraser asked the question, and didn't took the time to answer at that time).

Let me detail it.
When the `/{db}/_session` request is received by sync gateway, it is handled through the `sync-gateway/rest/handler.go` `checkAuth(...)` function.

As there is a OIDC provider which has options, the code block around line 300 is invoked :

	// If oidc enabled, check for bearer ID token
	if context.Options.OIDCOptions != nil {
		if token := h.getBearerToken(); token != "" {
			h.user, _, err = context.Authenticator().AuthenticateUntrustedJWT(token, context.OIDCProviders, h.getOIDCCallbackURL)
			if h.user == nil || err != nil {
				return base.HTTPErrorf(http.StatusUnauthorized, "Invalid login")
			}
			return nil
		}

Hence an Authenticator is created and its `AuthenticateUntrustedJWT(...)` function is called.
From what I understand, this function is declared in `sync-gateway/auth/autho.go`.


And somewhere in go-oidc, from what I understand, HS256 certificates are rejected because go-oidc can't find the public key used to encode the payload (which is normal, since there is none).

For a HS256 encoded token (which is in the logs), the signifiant portion of the sync gateway log will contain the following lines

	2017-06-07T11:44:27.924Z OIDC+: AuthenticateJWT called with token: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL2hvbWVib29rLmV1LmF1dGgwLmNvbS8iLCJzdWIiOiJhdXRoMHw1OTFlZmZmZDY2MGI4NzcxMTYxZTYwNDAiLCJhdWQiOiJ1OEM2QjVPV05LVUx2bmt1OTFCWXlLRXd5QmdSTlZLSCIsImV4cCI6MTQ5Njg3MTg2NywiaWF0IjoxNDk2ODM1ODY3fQ.YJozpM8KNDP6aTX_nKJX46vGUKk3_hrecIVpNjiPAvs
	2017-06-07T11:44:27.924Z OIDC+: JWT issuer: https://homebook.eu.auth0.com/, audiences: [u8C6B5OWNKULvnku91BYyKEwyBgRNVKH]
	2017-06-07T11:44:27.924Z OIDC+: Call GetProviderForIssuer w/ providers: map[Auth0:0xc4201026c0]
	2017-06-07T11:44:27.924Z OIDC+: GetProviderForIssuer with issuer: https://homebook.eu.auth0.com/, audiences: [u8C6B5OWNKULvnku91BYyKEwyBgRNVKH]
	2017-06-07T11:44:27.924Z OIDC+: Provider matches, returning
	2017-06-07T11:44:27.924Z OIDC+: Provider for issuer: &{JWTOptions:{ValidationKey:<nil> SigningMethod:0xc42016b910} Issuer:https://homebook.eu.auth0.com/ Register:true ClientID:0xc42016b8f0 ValidationKey:0xc42016b900 CallbackURL:0xc420427070 DisableSession:false Scope:[] IncludeAccessToken:false UserPrefix:auth0_ DiscoveryURI:https://homebook.eu.auth0.com/.well-known/openid-configuration DisableConfigValidation:false OIDCClient:0xc4201c25b0 OIDCClientOnce:{m:{state:0 sema:0} done:1} IsDefault:true Name:Auth0}
	2017-06-07T11:44:27.924Z OIDC+: Client &{0x1235720 0xc420cc2580 {u8C6B5OWNKULvnku91BYyKEwyBgRNVKH 1Y1psZ5_Ke3DAChYCxrIsvSfCQPAfqtAeZysF8-qwnFbaft8mMqj8iLOyRxEYZgn} http://localhost:4984/fx-example/_oidc_callback [openid email] {[{{NjgwNEJEODZEMEFEOTU3NzFGMDgyQTI4Q0FFREVFMUI4NzAxOUU1MQ RSA RS256 sig 65537 0xc42226c680 []}}] map[NjgwNEJEODZEMEFEOTU3NzFGMDgyQTI4Q0FFREVFMUI4NzAxOUU1MQ:0xc421af20e0] {63632519067 913536100 0x10cdc20}} <nil> {{0 0} 0 0 0 0} {63632432667 913547300 0x10cdc20}} could not verify JWT. Error: oidc: unable to verify JWT signature: no matching keys
	2017-06-07T11:44:27.924Z HTTP:  #382: POST /fx-example/_session
	2017-06-07T11:44:27.924Z HTTP: #382:     --> 401 Invalid login  (0.5 ms)

You can check in JWT that this token is an absolulty valid HS156 one. However, it fails ti validate in sync gateway go-oidc code. The point that lead me to understand it was due to algorithm is the following message : `Error: oidc: unable to verify JWT signature: no matching keys`. Why would sync gateway require a key to read a token which uses no key ?

**Comment by jamiltz** [comment]
Thanks for the details @Riduidel. I'd like to repro this but can't get Auth0 Open ID Connect to work with SG at the moment. I've created an API on Auth0 and using the Test client to get an ID token. It is valid when inserted in JWT.io. But when sent to Sync Gateway in the `/_session` endpoint I can see the following in the SG logs:

```
2017-06-07T16:41:12.527+01:00 OIDC+: JWT issuer: https://jamiltz.eu.auth0.com/, audiences: [http://sg.couchbase.com]
2017-06-07T16:41:12.527+01:00 OIDC+: Call GetProviderForIssuer w/ providers: map[Auth0:0xc42014c6c0]
2017-06-07T16:41:12.527+01:00 OIDC+: GetProviderForIssuer with issuer: https://jamiltz.eu.auth0.com/, audiences: [http://sg.couchbase.com]
2017-06-07T16:41:12.527+01:00 OIDC+: No provider match found
2017-06-07T16:41:12.527+01:00 OIDC+: Provider for issuer: <nil>
2017-06-07T16:41:12.527+01:00 HTTP:  #001: POST /default/_session
2017-06-07T16:41:12.527+01:00 HTTP: #001:     --> 401 Invalid login  (0.2 ms)
```

So there's something wrong in my setup. I don't think the `audiences: [http://sg.couchbase.com]` is right. I noticed in your forum thread that from your logs it should be `audiences: [CLIENT_ID]`.

Could you write the list of steps you took to get Auth0 OpenID Connect to work with SG, starting from Auth0's web UI? Thanks for your help!

**Pull request merged**